core/trino-main/src/main/java/io/trino/connector/DynamicCatalogManagerModule.java

```diff
@@ -36,17 +38,20 @@ protected void setup(Binder binder)
         if (buildConfigObject(ServerConfig.class).isCoordinator()) {
             binder.bind(CoordinatorDynamicCatalogManager.class).in(Scopes.SINGLETON);
             CatalogStoreConfig config = buildConfigObject(CatalogStoreConfig.class);
-            switch (config.getCatalogStoreKind().toLowerCase(Locale.ROOT)) {
-                case "memory" -> binder.bind(CatalogStore.class).to(InMemoryCatalogStore.class).in(Scopes.SINGLETON);
-                case "file" -> {
+            switch (config.getCatalogStoreKind()) {
+                case MEMORY -> binder.bind(CatalogStore.class).to(InMemoryCatalogStore.class).in(Scopes.SINGLETON);
+                case FILE -> {
                     configBinder(binder).bindConfig(FileCatalogStoreConfig.class);
                     binder.bind(CatalogStore.class).to(FileCatalogStore.class).in(Scopes.SINGLETON);
                 }
-                default -> {
+                case CUSTOM -> {
                     binder.bind(CatalogStoreManager.class).in(Scopes.SINGLETON);
-                    newOptionalBinder(binder, CatalogStoreManager.class).setBinding().to(CatalogStoreManager.class).in(Scopes.SINGLETON);
```

Comment:
This was incorrectly added here. The two solutions are:
- introduce an interface so that self-binding is not done
- remove `setBinding` so that Guice resolves the dependency from the object graph (already bound as a singleton just a line above)

From JavaDocs of `OptionalBinder` (https://google.github.io/guice/api-docs/4.2/javadoc/com/google/inject/multibindings/OptionalBinder.html) (emphasis mine).

> **_If neither setDefault nor setBinding are called, it will try to link to a user-supplied binding of the same type._** If no binding exists, the optionals will be absent. Otherwise, if a user-supplied binding of that type exists, or if setBinding or setDefault are called, the optionals will return present if they are bound to a non-null value.